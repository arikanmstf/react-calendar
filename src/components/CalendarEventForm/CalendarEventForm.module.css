.container {
    width: 209px;
    background-color: rgb(250, 250, 250);
    z-index: 99;
    border: 1px solid rgb(150, 150, 150);
    position: relative;
    padding: 8px;
    border-radius: 4px;
}

.header {
    display: flex;
    width: 100%;
}

.header > * {
    flex-grow: 1;
}

.content {
    display: flex;
    font-weight: bold;
    align-items: center;
    justify-content: space-between;
    margin-top: 8px;
}

.footer {
    margin-top: 48px;
    display: flex;
    align-items: center;
    flex-direction: row-reverse;
    justify-content: space-between;
}

.eventRow {
    display: flex;
    align-items: center;
    justify-content: space-between;
    border-bottom: 1px solid #ddd;
    padding: 4px 0;
}
