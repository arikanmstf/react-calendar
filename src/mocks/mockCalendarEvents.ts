const mockCalendarEvents = {
  1697760000000: [
    { title: "Mock event", date: new Date(1697760000000), id: "uuid-1" },
    { title: "Second mock event", date: new Date(1697760000000), id: "uuid-2" },
  ],
};

export default mockCalendarEvents;
