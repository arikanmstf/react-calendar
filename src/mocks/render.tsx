import { render as renderRTL } from "@testing-library/react";
import AppContextProvider from "../context/AppContextProvider";
import { ReactElement } from "react";

const render = (ui: ReactElement, { date }: { date?: Date } = {}) =>
  renderRTL(ui, {
    wrapper: ({ children }) => (
      <AppContextProvider date={date || new Date()}>
        {children}
      </AppContextProvider>
    ),
  });

export default render;
