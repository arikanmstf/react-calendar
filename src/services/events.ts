const EVENTS_KEY = "REACT_CALENDAR_APP_EVENTS";

// this typing is here, because probably this depends on backend response.
export type CalendarEvent = {
  date: number; // Date.valueOf
  title: string;
  id: string;
};

function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// GET /api/events
export const getEvents = async (): Promise<Record<number, CalendarEvent[]>> => {
  await sleep(300); // simulate api call.
  const events = window.sessionStorage?.getItem(EVENTS_KEY) || "{}";
  return JSON.parse(events);
};

// POST /api/events
export const postEvent = async (calendarEvent: CalendarEvent) => {
  if (window.sessionStorage) {
    // normally we don't need to fetch here at real api.
    const eventsObject = await getEvents();
    if (eventsObject[calendarEvent.date]) {
      eventsObject[calendarEvent.date].push(calendarEvent);
    } else {
      eventsObject[calendarEvent.date] = [calendarEvent];
    }
    window.sessionStorage.setItem(EVENTS_KEY, JSON.stringify(eventsObject));
  }
};

// DELETE /api/events/${id}
export const deleteEvent = async (calendarEvent: CalendarEvent) => {
  if (window.sessionStorage) {
    const eventsObject = await getEvents();
    if (eventsObject[calendarEvent.date]) {
      eventsObject[calendarEvent.date] = eventsObject[
        calendarEvent.date
      ].filter((event) => event.id !== calendarEvent.id);
      window.sessionStorage.setItem(EVENTS_KEY, JSON.stringify(eventsObject));
    }
  }
};

// delete all events
export const resetEvents = () => {
  window.sessionStorage?.setItem(EVENTS_KEY, "");
};
