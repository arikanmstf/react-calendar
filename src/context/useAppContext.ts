import { useContext } from "react";
import { getEvents } from "../services/events";
import { AppContext, AppDispatchContext } from "./AppContextProvider";

const useAppContext = () => {
  const context = useContext(AppContext);
  const dispatch = useContext(AppDispatchContext);

  const goToNextMonth = () => {
    if (context.monthsToShow.length < 3) {
      return;
    }
    dispatch({ type: "GO_TO_NEXT_MONTH" });
  };
  const goToPreviousMonth = () => {
    if (context.monthsToShow.length < 3) {
      return;
    }
    dispatch({ type: "GO_TO_PREVIOUS_MONTH" });
  };
  const jumpToDate = (date: Date) => {
    console.log("jumpToDatejumpToDate");
    dispatch({ type: "JUMP_TO_DATE", data: date });
  };

  const updateOpenPopoverIndex = (index: string) => {
    dispatch({ type: "UPDATE_OPEN_POPOVER_INDEX", data: index });
  };

  const fetchAndUpdateCalendarEvents = async () => {
    const events = await getEvents();
    dispatch({ type: "UPDATE_CALENDAR_EVENTS", data: events });
  };

  return {
    ...context,
    activeDate: context.monthsToShow[context.monthsToShow.length - 2],
    jumpToDate,
    goToNextMonth,
    goToPreviousMonth,
    updateOpenPopoverIndex,
    fetchAndUpdateCalendarEvents,
  };
};

export default useAppContext;
