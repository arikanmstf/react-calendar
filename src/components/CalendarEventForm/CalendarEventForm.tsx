import { v4 as uuidv4 } from "uuid";
import styles from "./CalendarEventForm.module.css";
import TextInput from "../../ui/TextInput/TextInput";
import { Bin, Calendar as CalendarIcon } from "../../ui/icons";
import { ChangeEvent, FC, FormEvent, useState } from "react";
import { Button } from "../../ui";
import useAppContext from "../../context/useAppContext";
import { CalendarEvent, deleteEvent, postEvent } from "../../services/events";
import { getEventsOfTheDate } from "../../utils/date";

type Props = {
  date: Date;
};

const CalendarEventForm: FC<Props> = ({ date }) => {
  const {
    updateOpenPopoverIndex,
    fetchAndUpdateCalendarEvents,
    calendarEvents,
  } = useAppContext();
  const [loading, setLoading] = useState(false);
  const [title, setTitle] = useState("");
  const eventsOnThisDay = getEventsOfTheDate(calendarEvents, date);

  const onTitleChange = (event: ChangeEvent<HTMLInputElement>) => {
    setTitle(event.target.value);
  };
  const onCancel = () => {
    updateOpenPopoverIndex("");
  };
  const onSubmit = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    setLoading(true);
    if (title) {
      await postEvent({ date: date.valueOf(), title, id: uuidv4() }); // BE should generate the ID
      await fetchAndUpdateCalendarEvents();
      setLoading(false);
      setTitle("");
    }
  };
  const onDelete = async (calendarEvent: CalendarEvent) => {
    setLoading(true);
    await deleteEvent(calendarEvent);
    await fetchAndUpdateCalendarEvents();
    setLoading(false);
  };

  return (
    <div className={styles.container}>
      <form onSubmit={onSubmit}>
        <div className={styles.header}>
          <TextInput
            value={title}
            onChange={onTitleChange}
            placeholder="Add new event"
            required
            maxLength={26}
          />
        </div>
        <div className={styles.content}>
          <CalendarIcon />
          <span>{date.toDateString()}</span>
        </div>
        <div className={styles.footer}>
          <Button disabled={loading} accent="POSITIVE" role="submit">
            Save
          </Button>
          <Button disabled={loading} accent="NEGATIVE" onClick={onCancel}>
            Cancel
          </Button>
        </div>
      </form>
      {eventsOnThisDay.length ? (
        <>
          <hr />
          <h5>Other events</h5>
          {eventsOnThisDay.map((calendarEvent, i) => (
            <div className={styles.eventRow} key={i}>
              <span>{calendarEvent.title}</span>
              <Button
                data-testid="delete-button"
                disabled={loading}
                onClick={() => onDelete(calendarEvent)}
              >
                <Bin />
              </Button>
            </div>
          ))}
        </>
      ) : null}
    </div>
  );
};

export default CalendarEventForm;
