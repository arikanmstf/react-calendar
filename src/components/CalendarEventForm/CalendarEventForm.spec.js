import React from "react";
import { screen, fireEvent } from "@testing-library/react";
import render from "../../mocks/render";
import CalendarEventForm from "./CalendarEventForm";
import { deleteEvent, postEvent } from "../../services/events";
import useAppContext from "../../context/useAppContext";
import mockCalendarEvents from "../../mocks/mockCalendarEvents";

jest.mock("../../context/useAppContext", () => jest.fn());

describe("CalendarEventForm", () => {
  beforeEach(() => {
    jest.resetAllMocks();

    useAppContext.mockReturnValue({
      updateOpenPopoverIndex: jest.fn(),
      fetchAndUpdateCalendarEvents: jest.fn(),
      calendarEvents: {},
    });
  });

  it("should render correctly", () => {
    const { container } = render(<CalendarEventForm date={new Date()} />);
    expect(screen.getByText("Fri Oct 20 2023")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("should prevent submitting with empty form", () => {
    render(<CalendarEventForm date={new Date()} />);
    fireEvent.click(screen.getByText("Save"));
    expect(postEvent).not.toHaveBeenCalled();
  });

  it("should make correct api call on submit", () => {
    render(<CalendarEventForm date={new Date()} />);
    fireEvent.change(screen.getByPlaceholderText("Add new event"), {
      target: {
        value: "My birthday",
      },
    });
    expect(screen.getByDisplayValue("My birthday")).toBeInTheDocument();
    fireEvent.click(screen.getByText("Save"));
    expect(postEvent).toHaveBeenCalledWith({
      date: 1697760000000, // value of 20/10/2023
      title: "My birthday",
      id: expect.any(String),
    });
  });

  describe("when listing events", () => {
    beforeEach(() => {
      useAppContext.mockReturnValue({
        updateOpenPopoverIndex: jest.fn(),
        fetchAndUpdateCalendarEvents: jest.fn(),
        calendarEvents: mockCalendarEvents,
      });
    });

    it("should list events correctly", () => {
      render(<CalendarEventForm date={new Date()} />);
      expect(
        screen.getByText(mockCalendarEvents["1697760000000"][0].title),
      ).toBeInTheDocument();
      expect(
        screen.getByText(mockCalendarEvents["1697760000000"][1].title),
      ).toBeInTheDocument();
    });

    it("should make correct api call on delete", () => {
      render(<CalendarEventForm date={new Date()} />);
      fireEvent.click(screen.getAllByTestId("delete-button")[0]);
      expect(deleteEvent).toHaveBeenCalledWith(
        mockCalendarEvents["1697760000000"][0],
      );
    });
  });
});
