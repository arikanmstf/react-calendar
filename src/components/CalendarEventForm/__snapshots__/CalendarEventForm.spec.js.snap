// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarEventForm should render correctly 1`] = `
<div>
  <div
    class="container"
  >
    <form>
      <div
        class="header"
      >
        <input
          class="textInput"
          maxlength="26"
          placeholder="Add new event"
          required=""
          value=""
        />
      </div>
      <div
        class="content"
      >
        <svg
          height="24"
          width="24"
        >
          calendar.svg
        </svg>
        <span>
          Fri Oct 20 2023
        </span>
      </div>
      <div
        class="footer"
      >
        <button
          class="button textButton positive  "
          role="submit"
        >
          Save
        </button>
        <button
          class="button textButton  negative "
        >
          Cancel
        </button>
      </div>
    </form>
  </div>
</div>
`;
